### DISCUSSION TITLE
什么时候该使用art渲染模版

### DISCUSSION THREAD

**Discussion by NeverBehave** [DIYgod/RSSHub#8359] [General]
在审核PR的过程中，发现一些可能在规范中并不清楚的地方，在这里说明一下art模版的想法

## 为什么用art

代码中嵌入大段模版，这会导致以下一些问题

1. 代码凌乱/过长，主要逻辑不清楚
2. 修改/替换不方便，和路由逻辑耦合

为了解决这个问题，我们开始推行大家使用art渲染html的意识

## 什么时候该用

简单来说：**任何，会最后输出到视图，且需要替换/拼接字符串的场景**

例如说：title标题，description描述，etc.

## 这样我要多学一个标记语言，太麻烦了

这其实就是webapp里面最基本的视图分离，把东西都放在路由内书写是最常见的anti-pattern
js可能处理起来确实方便，下面介绍一些可以让这个模版过程轻松一点的方法：

1. 将各个关键变量分离，如果最后需要输出的结果，小段拆出来，而不是整个合并在一起
2. 在模版中判断，并对应生成结果 —— 因为你需要的结果最后都会传入模版，判断其实速度很快，甚至你可以用js预处理一些if语句，这是可以接受的

如果还是不清楚在什么情况下应该使用模版：我们最近开放了`ctx.state.json`这个json输出渠道，那么任何按照常理，可以作为json api输出的结果，**应该被传入art模版渲染并得到最后的结果**

## 结语

其实这个模式我还在构思：对于每个模组，其实我们只需要
1. 提取的变量
2. 模版

那么这个渲染的过程其实可以由rsshub接管，同时我们也可以将模组可以获取的数据直接暴露出去。

___本文是为了解释为什么，我们现在审核路由过程中，不允许js中出现html；关于json化后续进程，可以在v2路由草案下继续讨论___

**Comment by tuzi3040** [comment]
> 同时，所有模版应该放在插件templates文件夹中 -- 后续我们会以此提供自定义模版切换/渲染等需求

能否大概描述一下 `自定义模版切换/渲染` 大概是个什么样子，使用不同的模板渲染时是否需要采用不同的 item.guid 让rss阅读器能够抓取已有 item 使用新模板渲染的内容（虽然会出现新旧内容重复

**Reply by NeverBehave** [reply]
这种相同内容不同模版至少不应该在同一个实例中出现：他应当被当作单独的rss源进行处理
